src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -68,10 +68,35 @@ void M(bool b)
             c.VerifyDiagnostics( );
         }
 
+        [Fact]
+        public void UnconstrainedAndErrorNullableFields()
+        {
+            var source = @"
+public class C<T>
+{
+    public T? field;
+    public Unknown? field2;
+}
+";
+
+            var c = CreateCompilation(new[] { source, NonNullTypesTrue, NonNullTypesAttributesDefinition });
+            c.VerifyDiagnostics(
+                // (5,12): error CS0246: The type or namespace name 'Unknown' could not be found (are you missing a using directive or an assembly reference?)
+                //     public Unknown? field2;
+                Diagnostic(ErrorCode.ERR_SingleTypeNameNotFound, "Unknown").WithArguments("Unknown").WithLocation(5, 12),
+                // (4,15): error CS8627: A nullable type parameter must be known to be a value type or non-nullable reference type. Consider adding a 'class', 'struct', or type constraint.
+                //     public T? field;
+                Diagnostic(ErrorCode.ERR_NullableUnconstrainedTypeParameter, "field").WithLocation(4, 15),
+                // (2,14): warning CS8618: Non-nullable field 'field' is uninitialized.
+                // public class C<T>
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "C").WithArguments("field", "field").WithLocation(2, 14)
```

Comment:
This warning feels a little off. Yes, they shouldn't have written `T?`, but they did, so I feel like we shouldn't be warning about the uninitialized field.

Reply:
I'm on the fence on that one. Most likely they mean for `T?` to be a nullable (either reference or value type), so we should warn.
I'm assuming that we'll produce this warning on nullable value types too.

---
In reply to: [215718737](https://github.com/dotnet/roslyn/pull/29636#discussion_r215718737) [](ancestors = 215718737)